### PR TITLE
add ruff for docstring enforcement and other linting/styling

### LINKED ISSUE #46: Additional Python linting capabilities
canu can catch stuff like this, so we probably need a ticket to add additional linting capabilities

_Originally posted by @jacobsalmela in https://github.com/Cray-HPE/libCSM/pull/44#discussion_r1202190904_

### PR DESCRIPTION
**Pull request by jacobsalmela**
### Summary and Scope

- Fixes #46 

#### Issue Type

- RFE Pull Request

### Prerequisites

- [x] I have included documentation in my PR (or it is not required).
- [x] I have added unit tests for my code.

### CODE REVIEW

File: libcsm/logger.py [hunk 1]
```diff
@@ -33,17 +31,16 @@
 
 
 class Logger(logging.Logger):
-    """
-    Inherits from logging.Logger, configuring custom settings on
-    initialization.
-    """
+    """Configure custom settings on initialization."""
```
Comment (rustydb) [inline libcsm/logger.py:34]: What's mandating the triple quotes on a single line, is that the `shortness`?

I _really_ don't like triple quotes on the same line.

**Review by rustydb — CHANGES_REQUESTED** [review]

File: libcsm/logger.py (hunk 1 above)
Comment (mitchty) [reply libcsm/logger.py:34]: Probably due to triple quote strings being docstrings so the newlines would end up in autogenerated docs.

**Review by mitchty — APPROVED** [review]

File: libcsm/logger.py (hunk 1 above)
Comment (jacobsalmela) [reply libcsm/logger.py:34]: I believe it comes from PEP https://peps.python.org/pep-0257/#one-line-docstrings. 

The strictness levels are:

- short (default): Expects a short summary and descriptions of arguments.
- long: Expects everything short expects, plus a longer description (if there is a newline in the docstring), and descriptions of return values and raised exceptions.
- full: Expects everything long expects, plus a description of yield values.

Comment (rustydb) [reply libcsm/logger.py:34]: `long` seems like it would like the new lines if I read that right..

Comment (rustydb) [reply libcsm/logger.py:34]: > Probably due to triple quote strings being docstrings so the newlines would end up in autogenerated docs.

For single-line things? I didn't notice anything weird in the sphinx docs https://github.com/Cray-HPE/libCSM/pull/37

Comment (jacobsalmela) [reply libcsm/logger.py:34]: `long` did not fix it.  I can just ignore D200.  Looks like the latest merges need a lot of fixing up though still to conform to the new linter

File: pyproject.toml [hunk 2]
```diff
@@ -91,4 +90,60 @@ requires = [
...
+
+[tool.ruff]
+# Enable pycodestyle (`E`), Pyflakes (`F`), docstrings (`D`)
+select = ["E", "F", "D"]
+ignore = [
+  "D107", # Missing docstring in `__init__`
+  "D200", # One-line docstring should fit on one line
+  "D203", # 1 blank line required before class docstring
+  "D205", # 1 blank line required between summary line and description
+  "D211", # No blank lines allowed before class docstring
```
Comment (rustydb) [inline pyproject.toml:103]: If this isn't ignored, does it then not care if there's a line or not?

File: pyproject.toml [hunk 3]
```diff
@@ -91,4 +90,60 @@ requires = [
...
+]
+
+# Same as Black.
+line-length = 88
+
+# Allow unused variables when underscore-prefixed.
+dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
+
+# Assume Python 3.11.
+target-version = "py311"
```
Comment (rustydb) [inline pyproject.toml:145]: Does this have to exist for ruff to work?

**Comment by rustydb** [comment]
Somehow pylint and ruff are not finding E302

My editor finds it, but ruff doesn't 
image

**Comment by rustydb** [comment]
> Somehow pylint and ruff are not finding E302
> 
> My editor finds it, but ruff doesn't image

if I run flake8 it finds it.

Not really clear on how we edit `selects` for ruff, it looks like it should be finding `^E` prefixed errors. Feels like it's skipping pep8 rules, which are most of what I care about.

**Comment by rustydb** [comment]
> > Somehow pylint and ruff are not finding E302
> > My editor finds it, but ruff doesn't image
> 
> if I run flake8 it finds it.
> 
> Not really clear on how we edit `selects` for ruff, it looks like it should be finding `^E` prefixed errors. Feels like it's skipping pep8 rules, which are most of what I care about and it seems we already were somehow skipping several of them.

Looks like it's not implemented yet.
https://github.com/astral-sh/ruff/issues/2402

**Review by rustydb — APPROVED** [review]

**Pull request merged**